### PR TITLE
[COMMIT] Eval code + scheduler + model load fn + dataset

### PR DESCRIPTION
**Pull request by justHungryMan**

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -48,32 +70,55 @@ def build_dataloader(self, ):
                 transforms.Normalize((0.1307, ), (0.3081, ))
             ]
         )
-        train_dataset = torchvision.datasets.MNIST(
+        train_dataset = torchvision.datasets.FashionMNIST(
```
Comment (justHungryMan) [inline train.py:73]: MNIST 쪽에 버그가 있어서 FashionMNIST 로 template 을 변경했습니다.

File: train.py [hunk 2]
```diff
@@ -91,37 +136,55 @@ def train_one_epoch(self, epoch, model, dl, criterion, optimizer):
...
-        for epoch in range(self.conf.hyperparameter.epochs):
-            sampler.set_epoch(epoch)
+        for epoch in range(1, self.conf.hyperparameter.epochs + 1):
+            train_sampler.set_epoch(epoch)
 
             # train
-            model.train()
-            train_loss, train_acc = self.train_one_epoch(epoch, model, dl, criterion, optimizer)
-            
+            train_loss, train_acc = self.train_one_epoch(epoch, model, train_dl, criterion, optimizer)
```
Comment (justHungryMan) [inline train.py:181]: summarywriter or wandb 같이 visualization tool 이 필요합니다.

File: trainer/dataset/__init__.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    
+    sampler = torch.utils.data.distributed.DistributedSampler(
+            dataset, 
+            num_replicas=world_size, 
+            rank=local_rank,
+            shuffle=(mode == 'train')
+        )
+    dl = torch.utils.data.DataLoader(
+        dataset,
+        batch_size=conf['batch_size'],
```
Comment (LeeChanHyuk) [inline trainer/dataset/__init__.py:32]: GPU 개수에 따라서 각자 따로 계산해서 설정할수도 있겠지만 batch_size를 GPU 개수로 나눠주면 어떨까요?

File: trainer/dataset/__init__.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        shuffle=False,
+        pin_memory=True,
+        drop_last=conf['drop_last'],
+        num_workers=os.cpu_count(),
```
Comment (LeeChanHyuk) [inline trainer/dataset/__init__.py:36]: 개인적으로는 머신에 따라서 num_workers 오류가 났던적이 제법 있어서 os.cpu_count()보다는 gpu 개수 * 4 정도로 추천받아서 써봤을 때 좋았던 기억이 있습니다.

**Review by justHungryMan — COMMENTED** [review]
.

File: trainer/dataset/__init__.py (hunk 3 above)
Comment (justHungryMan) [reply trainer/dataset/__init__.py:32]: 음.. 왜 쓰신 글 밑에 쓰레드가 안달리는지 모르겠네요.
@LeeChanHyuk 이렇게 한 이유는 각자 실험 환경이 다를텐데 (gpu 4개, gpu 2개 등등..) 다른 환경에서 최대한 같은 실험결과를 내기 위해서 single process per single gpu 로 하였습니다.

**Review by LeeChanHyuk — APPROVED** [review]

File: trainer/dataset/__init__.py (hunk 4 above)
Comment (justHungryMan) [reply trainer/dataset/__init__.py:36]: 그럼 num_workers=4 로 하면 될까요?

Comment (LeeChanHyuk) [reply trainer/dataset/__init__.py:36]: 넵ㅎㅎ 좋습니다

File: conf/hyperparameter/toy.yaml [hunk 5]
```diff
@@ -1,2 +1,3 @@
 batch_size: 64
```
Comment (jhb115) [inline conf/hyperparameter/toy.yaml:1]: dataset/mnist.py에도 batch_size가 정의되어 있어 서로 중복되는 것 같아요. 하나로 통일하면 좋을 것 같습니다

File: train.py [hunk 6]
```diff
@@ -31,97 +34,158 @@ def set_env(self):
...
+
+        return train_loader, train_sampler, test_loader, test_sampler
 
     def build_loss(self, ):
         criterion = trainer.loss.create(self.conf.loss, self.rank)
+        criterion.to(device=self.rank, non_blocking=True)
 
         return criterion
+    
+    def load_model(self, model, path):
```
Comment (jhb115) [inline train.py:101]: 추후에 save_model 함수 구현도 필요하겠네요.

File: train.py [hunk 7]
```diff
@@ -31,97 +34,158 @@ def set_env(self):
...
+        if self.conf.scheduler.params.get('T_max', None) is None:
+            self.conf.scheduler.params.T_max = self.conf.hyperparameter.epochs
+        
+        # warmup Scheduler
+        # if self.conf.scheduler.warmup.get('status', False) is True:
+        #     self.conf.scheduler.warmup.params.total_epoch = self.conf.hyperparameter.epochs
     
     def build_model(self, num_classes=-1):
-        model = trainer.architecture.create(self.conf.architecture, rank=self.rank)
+        model = trainer.architecture.create(self.conf.architecture)
```
Comment (jhb115) [inline train.py:61]: trainer.architecture.create 함수에서 pre-trained된 모델을 불러오는 기능을 포함하거나, 여기 build_model 함수에서 불러오는 기능을 구현하면 좋을 것 같습니다.

File: train.py [hunk 8]
```diff
@@ -31,97 +34,158 @@ def set_env(self):
...
+        criterion.to(device=self.rank, non_blocking=True)
 
         return criterion
+    
+    def load_model(self, model, path):
+        data = torch.load(path)
+        key = 'model' if 'model' in data else 'state_dict'
+
+        if not isinstance(model, (DataParallel, DDP)):
+            model.load_state_dict({k.replace('module.', ''): v for k, v in data[key].items()})
```
Comment (jhb115) [inline train.py:106]: 흠... optimizer의 load state도 필요할까요? 만약에 도중에 중단된 학습을 재개하고 싶을 경우에 필요할 것 같습니다.

File: train.py [hunk 9]
```diff
@@ -31,97 +34,158 @@ def set_env(self):
...
-        dl, sampler = self.build_dataloader()
+        scheduler = self.build_scheduler(optimizer)
+        train_dl, train_sampler, test_dl, _ = self.build_dataloader()
 
         # Wrap the model
-        model = nn.parallel.DistributedDataParallel(model, device_ids=[self.rank])
         
         # initialize
         for name, x in model.state_dict().items():
             dist.broadcast(x, 0)
```
Comment (jhb115) [inline train.py:175]: 이 부분은 무슨 역할인가요?

Comment (jhb115) [reply train.py:175]: 마스터 노드에 있는 모델을 다른 노드들로 broadcast해주는 것 같은데, build_model()함수의 DDP가 자동적으로 뿌려주지는 않나요?

**Review by jhb115 — APPROVED** [review]

File: conf/hyperparameter/toy.yaml (hunk 5 above)
Comment (justHungryMan) [reply conf/hyperparameter/toy.yaml:1]: 이부분은 dataloader 쪽에서 batch_size 를 conf/dataset/{dataname}.yaml 에서 batch_size: ??? 로 정의되어있는데 이부분을 채우기 위하여 https://github.com/Psudo-kaggle/kaggle-template/blob/f35768528173f2b8be9255247e0850ba80b2b88a/train.py#L50
여기서 ??? 인 부분을 hyperparameter 의 batch_size 로 채워주고 있습니다.

이렇게 한 이유는 각 실험환경의 gpu 가 다를텐데 (memory 등등) 거기에 따라 batch_size 가 달라지기 때문에 hyperparameter config 에서 컨트롤하려고 하였습니다.
conf/dataset/{dataname}.yaml 은 건들이지 말구요.

다른 방법으로는 conf/hyperparameter 의 batch_size 를 conf/dataset/{datasetname} 에 등록해주지 말고 parameter 로 함수에 넘기는 방법이 있는데 저는 이 방법을 선호해서 현재와 같이 코드작업을 하였습니다.

이 부분에 대해서 추가 의견 및 다른분들의 의견을 듣고싶습니다.
개선 가능하다면 좋은 코드로 바꿨으면 좋겠네요.
@jhb115

File: train.py (hunk 6 above)
Comment (justHungryMan) [reply train.py:101]: 네 여기를 참고해서 save_model 을 관리하면 좋겠다고 생각하고 있습니다.
https://github.com/rwightman/pytorch-image-models/blob/acd6c687fd1c0507128f0ce091829b233c8560b9/timm/utils/checkpoint_saver.py#L21

File: train.py (hunk 8 above)
Comment (justHungryMan) [reply train.py:106]: 넵 이것도 위의 코멘트에 첨부한 saver 모듈을 확장해서 관리 가능할 것 같은데요, 파이토치는 학습재개 checkpoint 를 load 해도 정확하게 load가 안된다는 보고가 있는데 다른분들 아시는 사항 있으신가요?

File: train.py (hunk 9 above)
Comment (justHungryMan) [reply train.py:175]: https://github.com/Psudo-kaggle/kaggle-template/blob/f35768528173f2b8be9255247e0850ba80b2b88a/train.py#L174
여기 말씀하시는거죠?

제가 듣기로는 처음 모델이 initialize 된 상태에서는 모델의 weight 들이 각 process node (one process per gpu node) 마다 다르다고 들었습니다. (이게 개선되었는지는 모르겠네요.)
그래서 초기에 model 을 정의해주고 0번째 gpu devcie 의 model weight 를 전체 gpu node 들의 model weight 로 바꿔주는 기능을 합니다.

File: train.py (hunk 2 above)
Comment (pcmin03) [reply train.py:182]: 넵 새로 branch를 파서 작업중에 있습니다. 
빠른 시일안에 pL을 하도록 하겠습니다 (회사일이 정신없네요..ㅎ)

**Review by pcmin03 — APPROVED** [review]
